File: 1751 (Maximum Number of Events That Can Be Attended II)/python/Solution.py
class Solution:
    def maxValue(self, events: List[List[int]], k: int) -> int:
        events.sort()
        starts = [s for s,e,v in events]

        @lru_cache
        def solve(idx, k):
            if idx>=len(events) or k==0:
                return 0
            next_event_idx=  bisect.bisect_right(starts,events[idx][1])
            return max(solve(idx+1,k), events[idx][2] + solve(next_event_idx,k-1))

        return solve(0,k)

File: 1751 (Maximum Number of Events That Can Be Attended II)/cpp/Solution.cpp
class Solution {
public:

    int solve(int idx, int k , vector<vector<int>> &events, vector<int> &starts, vector<vector<int>> &dp){
        if (k==0 || idx>=events.size()) return 0;
        if (dp[idx][k]!=-1) return dp[idx][k];
        int new_idx = std::upper_bound(starts.begin(), starts.end(), events[idx][1]) - starts.begin();
        dp[idx][k] = max(solve(idx+1,k,events,starts,dp), events[idx][2]+ solve(new_idx,k-1,events,starts,dp));
        return dp[idx][k];
    }
    int maxValue(vector<vector<int>>& events, int k) {
        sort(events.begin(),events.end());
        vector<int> starts;
        for(int i = 0;i<events.size();i++){
            starts.push_back(events[i][0]);
        }
        vector<vector<int>> dp(events.size(), vector<int>(k+1,-1));
        return solve(0,k,events,starts,dp);
    }
};